### ISSUE TITLE
The discriminator of the adversarial network.

### ISSUE THREAD

**Issue by Renp1ngs** [open-mmlab/mmsegmentation#2110]
Hi!
I want to implement a discriminator in segmentor. In addition, I want to implement an optimizer for this discriminator alone, how can I quickly build this optimizer and lr_config.

**Comment by MengzhangLI** [comment]
Hi, first you could create a `discriminator` class together with backbone, neck and head. Then you could follow the total framework of [segmentor](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/segmentors/encoder_decoder.py#L14) and adding it in a proper way. Then you could follow [`train_segmentor`](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/apis/train.py#L120-L121) about how to build runner.

Plus, you can follow MMGeneration codebase which has mature framework of discriminator and [generator](https://github.com/open-mmlab/mmgeneration/blob/adae47db96cc7cf873ad2b143a21a6bbd27d3c10/mmgen/models/architectures/cyclegan/generator_discriminator.py).

**Comment by Yongze** [comment]
Hi
I also add a discriminator in the framework. But I have a problem: in the gradient reversal part, only forward() gets called, and backward() does not get called. The code is like the one below. BTW, I added GradientReversal() in the first layer of decoder.

`# https://github.com/janfreyberg/pytorch-revgrad
from torch.autograd import Function
from torch.nn import Module
from torch import tensor


class RevGrad(Function):
    @staticmethod
    def forward(ctx, input_, alpha_):
        ctx.save_for_backward(input_, alpha_)
        output = [i.clone() for i in input_]
        return output

    @staticmethod
    def backward(ctx, grad_output):  # pragma: no cover
        grad_input = None
        _, alpha_ = ctx.saved_tensors
        if ctx.needs_input_grad[0]:
            grad_input = -grad_output * alpha_
        return grad_input, None


class GradientReversal(Module):
    def __init__(self, alpha=1., *args, **kwargs):
        """
        A gradient reversal layer.
        This layer has no parameters, and simply reverses the gradient
        in the backward pass.
        """
        super().__init__(*args, **kwargs)

        self._alpha = tensor(alpha, requires_grad=False)

    def forward(self, input_):
        return RevGrad.apply(input_, self._alpha)
`
